### PR TITLE
Fix syncing calico node labels for Kubernetes CRD mode

### PR DESCRIPTION
**Pull request by huiyizzz**
## Description

For Kubernetes CRD mode, sync labels should always be set to enabled.
This PR adds a validation rule for `syncLabels`. When the datastore type is Kubernetes, it does not allow disabling sync labels.
## Related issues/PRs

https://github.com/projectcalico/calico/issues/5594
## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Fixed SyncLabels validation for Kubernetes datastore.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: api/go.sum [hunk 1]
```diff
@@ -166,6 +166,7 @@ github.com/google/pprof v0.0.0-20201203190320-1bf35d6f28c2/go.mod h1:kpwsk12EmLe
 github.com/google/pprof v0.0.0-20210122040257-d980be63207e/go.mod h1:kpwsk12EmLew5upagYY7GY0pfYCcupk39gWOCRROcvE=
 github.com/google/pprof v0.0.0-20210226084205-cbba55b83ad5/go.mod h1:kpwsk12EmLew5upagYY7GY0pfYCcupk39gWOCRROcvE=
 github.com/google/renameio v0.1.0/go.mod h1:KWCgfxg9yswjAJkECMjeO8J8rahYeXnNhOm40UhjYkI=
+github.com/google/uuid v1.1.2 h1:EVhdT+1Kseyi1/pUmXKaFxYsDNy9RQYkMWRH68J/W7Y=
```
Comment (coutinhop) [inline api/go.sum:169]: The changes to this file don't seem related to your other changes, are they necessary for this PR? (or are they just from running some `go mod` command?)

Could you try reverting these and testing if everything works without this file's changes?

File: libcalico-go/lib/validator/v3/validator.go [hunk 2]
```diff
@@ -445,6 +447,17 @@ func validateLabels(fl validator.FieldLevel) bool {
 	return true
 }
 
+func validateSyncLabels(fl validator.FieldLevel) bool {
+	s := fl.Field().String()
+	log.Debugf("Validate SyncLabels: %s", s)
+	dst := os.Getenv("DATASTORE_TYPE")
+	if s == "Disabled" && dst == "kubernetes" {
```
Comment (coutinhop) [inline libcalico-go/lib/validator/v3/validator.go:454]: I'm not sure if it would be too extreme (@mgleung let me know if it is), but could we maybe use `strings.ToLower()` in order to still catch cases where, for example, DATASTORE_TYPE is "Kubernetes"?
(if we do go ahead with this, please remember to add test cases for it in validator_test.go)

**Review by coutinhop — CHANGES_REQUESTED** [review]
Overall LGTM, just a couple of questions/comments...

Thanks!

File: libcalico-go/lib/validator/v3/validator.go (hunk 2 above)
Comment (mgleung) [reply libcalico-go/lib/validator/v3/validator.go:454]: I just had a chat with @huiyizzz and we'll try a different approach. I don't think that we can guarantee that the env variable will be set each time for every component that calls this so we'll try something that doesn't depend upon this value.

Comment (mgleung) [reply libcalico-go/lib/validator/v3/validator.go:454]: `strings.ToLower()` would be great to use here though normally to ensure we're matching properly though.

File: libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go [hunk 3]
```diff
@@ -30,7 +43,7 @@ const (
 )
 
 func NewKubeControllersConfigClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sResourceClient {
-	return &customK8sResourceClient{
+	return &kubeControllersConfigClient{
```
Comment (caseydavenport) [inline libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go:46]: @huiyizzz rather than reimplement all of `customK8sResourceClient`, would it be possible to add a new option field to it?

e.g.,

```go
type customK8sResourceClient struct {
  validator Validator
}

// A Validator is a function that validates a resource and returns an error if it is not valid.
type Validator func(kvp KVPair) error
```

Then the Create / Update methods can have something like this in it:

```go
if c.Validator != nil {
  if err = c.Validator(kvp); err != nil {
    return nil, err
  }
}
```

WDYT?

Comment (coutinhop) [reply libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go:46]: +1 (I was just writing a comment on how we should try to avoid "copying" `customK8sResourceClient`, @caseydavenport's solution is better than what I had)

Comment (huiyizzz) [reply libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go:46]: Oh right! Sounds like a better solution!

File: libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go [hunk 4]
```diff
@@ -17,11 +17,13 @@ package resources
...
 
-	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
-
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	v3 "github.com/projectcalico/calico/libcalico-go/lib/validator/v3"
```
Comment (caseydavenport) [inline libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go:26]: I'd recommend calling this import "validator" - "v3" is too vague and normally refers to the api/projectcalico/v3 package in other locations, so this might be confusing.

File: libcalico-go/lib/validator/v3/validator.go [hunk 5]
```diff
@@ -159,6 +159,7 @@ func init() {
 	registerFieldValidator("containerID", validateContainerID)
 	registerFieldValidator("selector", validateSelector)
 	registerFieldValidator("labels", validateLabels)
+
```
Comment (coutinhop) [inline libcalico-go/lib/validator/v3/validator.go:162]: nit: no need to add newline here

File: libcalico-go/lib/validator/v3/validator_test.go [hunk 6]
```diff
@@ -2723,25 +2723,6 @@ func init() {
 			Node:    "node-1",
 		}, false),
 	)
-
-	DescribeTable("SyncLabels Validator",
```
Comment (coutinhop) [inline libcalico-go/lib/validator/v3/validator_test.go]: I think we need to have an equivalent test now that the code was moved to libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go, perhaps add a kubecontrollersconfig_test.go if we don't have one already?

**Review by coutinhop — APPROVED** [review]

File: libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go [hunk 7]
```diff
@@ -17,18 +17,23 @@ package resources
...
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
 )
 
 const (
 	KubeControllersConfigResourceName = "KubeControllersConfigurations"
 	KubeControllersConfigCRDName      = "kubecontrollersconfigurations.crd.projectcalico.org"
 )
 
+var validate *validator.Validate
```
Comment (mgleung) [inline libcalico-go/lib/backend/k8s/resources/kubecontrollersconfig.go:35]: Though I like the pattern of adding a validator object in order to make it extensible to other resources defined by the `customResourceClient`, I think that the validator we use here can be much simpler than the validator we use for the client in general. I think it might be easier (and a little cleaner) to define a separate validator here that will just have a `Validate` method that is overridden per resource. So in our `KubeControllersConfig` case, `Validate` would only need to validate the `k8sSyncLabels` field and we wouldn't need to use an `init` call or a separate call to `Var` in order to target the specific fields to validate.

**Review by mgleung — APPROVED** [review]
LGTM
